### PR TITLE
NeuralFoil Synthetic Data Flooding: single-foil Cp augmentation via neural surrogate

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The training set has limited coverage of the (AoA, Re) parameter space for single-foil configurations. By using NeuralFoil (Sharpe, 2024, arXiv:2503.16323) — a neural surrogate trained on 1M+ XFoil runs that predicts full Cp distributions in <1ms — we can generate thousands of additional synthetic single-foil surface pressure samples spanning a dense grid of AoA, Re, thickness, and camber values.

This directly expands the training distribution rather than adding input features (which failed in prior panel-method experiments). The synthetic samples provide surface-only supervision — volume nodes are masked during training for these samples.

**Key bet:** Broader single-foil coverage improves in-distribution precision (p_in) and OOD generalization (p_oodc, p_re) by giving the model more diverse surface pressure patterns to learn from. This is a DATA GENERATION approach per human directive #1860.

**Literature:**
- Sharpe "NeuralFoil" (arXiv:2503.16323, 2024) — <1ms per query, competitive with XFoil
- Vinuesa & Brunton (Nature Comp Sci, 2022) — synthetic data flooding for CFD surrogates

### Phase 1: Offline Data Generation (run once before training)

Create a script `cfd_tandemfoil/generate_neuralfoil_data.py`:

1. Install NeuralFoil: `pip install neuralfoil`
2. Generate 5000 synthetic single-foil samples:
   - AoA: uniform(-15, 20) degrees
   - Re: uniform(5e4, 2e6)
   - NACA thickness: uniform(0.08, 0.20)
   - NACA camber: uniform(0.0, 0.06)
3. For each sample, call `neuralfoil.get_aero_from_airfoil_name()` with `model_size="large"` to get surface Cp distributions (Cp_upper, Cp_lower at chord-wise positions)
4. Convert Cp to pressure: `p = p_inf + 0.5 * rho * V_inf^2 * Cp` (use Re to derive V_inf with assumed chord=1, kinematic viscosity=1.5e-5)
5. Save to a directory `data/neuralfoil_synthetic/` as individual .pt files with fields: `aoa`, `re`, `surface_xy`, `surface_p`, `surface_upper_mask`
6. Interpolate NeuralFoil's chord-wise Cp points onto the TandemFoilSet single-foil surface node x-positions via linear interpolation

### Phase 2: Training Integration in train.py

1. Add flag `--neuralfoil_synthetic_path` (str, default None) pointing to the generated data directory
2. Add flag `--p_synthetic 0.3` (float) — probability of sampling a synthetic sample in each batch position
3. When loading a synthetic sample:
   - Surface nodes: use the NeuralFoil-derived pressure as target
   - Volume nodes: set domain_mask=0 (no volume supervision for synthetic samples)
   - Input features: construct the standard 24-dim features from the synthetic geometry (DSDF from surface coords, domain flags, Re, AoA)
   - PCGrad: assign synthetic samples to the `single_foil` gradient task group
4. The synthetic DataLoader should be separate from the main TandemFoilSet DataLoader, with random mixing at the batch level

### Phase 3: Training

Run 2 seeds with the full baseline config plus synthetic data:

```
cd cfd_tandemfoil && python train.py --agent alphonse --seed 42 \
  --wandb_name "alphonse/neuralfoil-synthetic-flood-s42" \
  --wandb_group "neuralfoil-synthetic-flood" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --neuralfoil_synthetic_path data/neuralfoil_synthetic --p_synthetic 0.3
```

Then seed 73 with same flags but `--seed 73`.

### Important Implementation Notes

- NeuralFoil outputs Cp (pressure coefficient), NOT raw pressure. You MUST convert to raw pressure using the dynamic pressure formula.
- The asinh transform applies to targets — apply it to synthetic targets too.
- Validate by comparing NeuralFoil Cp at a few AoA/Re values against actual TandemFoilSet single-foil samples before running full training.
- If TandemFoilSet uses non-NACA geometries, check `data/README.md` — NeuralFoil is trained on NACA-family airfoils.
- Run the data generation script first, verify it produces valid samples, then proceed to training.

## Baseline

Current best metrics (PR #2251, 2-seed avg):
- p_in: **11.891** (target: < 11.89)
- p_oodc: **7.561** (target: < 7.56)
- p_tan: **28.118** (target: < 28.12)
- p_re: **6.364** (target: < 6.36)
- W&B baseline runs: 7jix2jkg (s42), epkfhxfl (s73)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT alphonse:

## Results

### Training command (seed 42)
\`\`\`bash
cd cfd_tandemfoil && python train.py --agent alphonse --seed 42 \
  --wandb_name "alphonse/neuralfoil-synthetic-flood-s42" \
  --wandb_group "neuralfoil-synthetic-flood" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --neuralfoil_synthetic_path data/neuralfoil_synthetic --p_synthetic 0.3
\`\`\`

Seed 73 same command with `--seed 73` and adjusted `--wandb_name`.

### Metrics

| Metric | s42 | s73 | 2-seed avg | baseline (PR #2251) | Δ |
|--------|-----|-----|-----------|---------------------|---|
| **p_in** | 13.76 | 13.77 | **13.77** | 11.891 | **+15.8%** ⛔ |
| **p_tan** | 28.6 | 29.7 | **29.15** | 28.118 | **+3.7%** ⛔ |
| **p_oodc** | 9.5 | 9.4 | **9.45** | 7.561 | **+24.9%** ⛔ |
| **p_re** | 7.3 | 7.2 | **7.25** | 6.364 | **+13.9%** ⛔ |

Surface MAE Ux/Uy (val_in_dist): s42 = 3.60/1.10, s73 = 3.43/1.04

### W&B run IDs
- Seed 42: `k2rhfngo` (alphonse/neuralfoil-synthetic-flood-s42)
- Seed 73: `sunnwbxy` (alphonse/neuralfoil-synthetic-flood-s73)

### Peak memory
- Seed 42: 44.9 GB (vs baseline ~41 GB — +4 GB from synthetic dataset in RAM)
- Seed 73: 45.8 GB

### Epochs completed
- Seed 42: 157 epochs (180.1 min wall clock)
- Seed 73: 158 epochs (181.2 min wall clock)

### What happened — negative result

NeuralFoil synthetic data flooding **significantly hurts all metrics** (p_in +15.8%, p_oodc +24.9%, p_re +13.9%, p_tan +3.7%). This is a clear failure.

**Primary failure mode — geometric inconsistency**: The synthetic data is generated by taking 20 template meshes (real single-foil CFD samples) and assigning NeuralFoil Cp values for a *randomly sampled NACA* at a random (AoA, Re). Crucially, the node positions (x, y) and all derived geometric features (SDF, curvature, arc-fraction, DSDF) remain those of the **template** NACA, while `x_syn[:, 15:18]` (NACA encoding) is updated to the randomly chosen NACA. The model therefore sees a contradictory training signal: "this node is geometrically located at position consistent with NACA-A, but the NACA encoding says NACA-B, and the pressure is what you'd see on NACA-B at this AoA." 

For example, a surface node from a NACA0012 template at (x=0.4, y=0.06) gets the Cp of NACA2412 at x=0.4 — but the NACA0012 surface at x=0.4 is at a different y-coordinate than NACA2412. The node's geometric context (SAF, DSDF, curvature) says NACA0012, the encoding says NACA2412, and the target pressure is a hybrid. The model cannot resolve this and learns spurious geometry-pressure correlations that conflict with real CFD data.

**Secondary factors:**
- NeuralFoil is a panel-method surrogate; its Cp errors vs OpenFOAM CFD may be non-trivial near stagnation/leading edge, introducing systematic noise
- With p_synthetic=0.3, 30% of training slots are noisy synthetic samples, substantially diluting real data signal
- The 4803 synthetic samples all share only 20 geometric template meshes — very low geometric diversity for the fraction of training they occupy

### Suggested follow-ups

1. **Geometry-consistent synthesis**: Generate synthetic data by *fixing the NACA* to match the template mesh. Only vary AoA and Re for that same NACA geometry. This eliminates the geometry/encoding mismatch — the model sees consistent (position, NACA, Cp) triples. The coverage gain is smaller (only (AoA, Re) space, not NACA space) but the data would be valid.

2. **Direct CFD augmentation of AoA**: Instead of NeuralFoil, run a small batch of cheap RANS simulations on existing mesh geometries at perturbed AoA values (±5°). This keeps geometry exactly consistent.

3. **Surface-only loss but with smaller p_synthetic**: Try p_synthetic=0.05 or 0.1. Even with the geometry mismatch, a smaller fraction of noisy samples might add some regularization benefit rather than dominating the loss.

4. **Use NeuralFoil only for data augmentation of existing geometries**: For each real training sample, also generate a NeuralFoil-derived version at the same AoA/Re but perturbed by ±2°. This is geometry-consistent (same node positions, same NACA encoding) and tests whether NeuralFoil-quality targets can help for AoA perturbation augmentation.

### Pre-existing bug note
The post-training verification step reports NaN for Manual denorm MAE surf_p on the OOD-Re split. This appears to be a pre-existing issue unrelated to this PR (present in both seeds).

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2275 because ADVISOR: Closing — all 4 metrics regressed significantly (p_in +15.8%, p_oodc +24.9%, p_tan +3.7%, p_re +13.9%).

**Root cause — geometric inconsistency**: Synthetic samples use template mesh node positions (geometry of NACA-A) but assign NeuralFoil Cp for a different NACA-B. The model sees contradictory geometry-pressure signals — node coordinates say one airfoil shape, NACA encoding says another, and target pressure is for the encoded (not geometric) shape. This is an irreconcilable conflict that corrupts learning.

**Secondary factors**: NeuralFoil panel-method Cp errors vs OpenFOAM near stagnation, 30% synthetic diluting real data, only 20 template meshes for low geometric diversity.

Good analysis of the failure mode and suggested follow-ups. The geometry-consistent variant (fix NACA to match template, vary only AoA/Re) is theoretically sound but the coverage gain would be small. Moving you to a fresh direction..
